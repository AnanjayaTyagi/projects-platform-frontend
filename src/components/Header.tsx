import * as React from "react";
import { Link, NavLink } from "react-router-dom";
import { AppContext } from "../contexts/AppContext";
import { User } from "../types";
import "../style/components/Header.css";
import axios from "axios";
import Cookies from 'js-cookie';

type Props = {
  user?: User;
};

type State = {
  is_authenticated: boolean;
}

class Header extends React.Component<Props, State> {
  static contextType = AppContext;

  constructor(props: Props) {
    super(props);
    this.signUserOut = this.signUserOut.bind(this);
    this.signUserIn = this.signUserIn.bind(this);
    this.state = { is_authenticated: !!Cookies.get("auth_token") };

    // this.logout = this.logout.bind(this);
  }

  signUserIn(e: any) {
    window.location.href = "http://127.0.0.1:8000/accounts/github/login";
  }

  async signUserOut(e: any) {
    console.log("xx logging out", this.context.user.auth_token);
    const result = await axios(
      { method: 'get', url: "http://127.0.0.1:8000/projects/logout/", headers: { Authorization: `Token ${this.context.user.auth_token}` } }
    );
    console.log("logout result", result);
    if (result.data['result'] === 'success') {
      Cookies.remove('auth_token', { path: '/' });
      console.log("signUserOut after remove VVVVV cookies", Cookies.get());
      this.context.user = {
        name: "",
        is_authenticated: false,
        csrf_token: "",
        auth_token: "",
      };
      this.setState({ is_authenticated: false });
      console.log("XXXXXX removed auth_token from cookies and reset user", Cookies.get(), this.context.user);

    }
  }

  // logout(e: any) {
  //   this.signUserOut(e).then(() => this.render());
  // }

  componentDidMount() {

    console.log("componentDidMount user.is_authenticated", !!Cookies.get("auth_token"));
    this.setState({ is_authenticated: !!Cookies.get("auth_token") });
    console.log("state is_authenticated", this.state.is_authenticated);
  }

  render() {
    const user = this.context.user;
    const signedOut = (
      <div className="auth">
        <button onClick={this.signUserIn}>Sign in with GitHub</button>
      </div>
    );

    const signedIn = (
      <div className="auth">
        <p>Hello, {user.name}!</p>
        <Link to="add-project" className="button-link">Add Project</Link>
        <button onClick={this.signUserOut}>Log Out</button>
      </div>
    );

    return (
      <header>
        <h1 className="logo">
          <Link to="/projects">
            she's collaborating
            </Link>
        </h1>
        <nav>
          <ul>
            <li><NavLink exact to="/" activeClassName="current">Home</NavLink></li>
            <li><NavLink to="/projects" activeClassName="current">Projects</NavLink></li>
          </ul>
        </nav>
        {this.state.is_authenticated ? signedIn : signedOut}
      </header>
    );
  }
}

export default Header;
