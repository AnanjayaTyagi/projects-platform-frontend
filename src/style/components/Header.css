header {
  padding: 40px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #FFF;
  /* border-bottom: 1px dotted #d1d1d3; */
}
.logo {
  font-size: 1.4em;
  border-bottom: 3px solid #FFAD91;
}
.auth {
  display: flex;
  align-items: center;
}
.auth > * {
  margin-left: 24px;
}

header button {
  font-size: 18px;
  padding: 12px 18px;
  color: #fff;
  background-color: #FFAD91;
  border: none;
  border-radius: 36px;
}
