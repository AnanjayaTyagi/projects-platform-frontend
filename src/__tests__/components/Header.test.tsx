import * as React from "react";
import renderer from "react-test-renderer";

import Header from "../../components/Header";

it("renders Header correctly", () => {
  const component = renderer.create(<Header />).toJSON();
  expect(component).toMatchSnapshot();
});
