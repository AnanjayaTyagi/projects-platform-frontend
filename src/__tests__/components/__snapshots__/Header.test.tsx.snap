// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Header correctly 1`] = `
<header>
  <h1
    className="logo"
  >
    she's collaborating
  </h1>
  <div
    className="auth"
  >
    <button
      onClick={[Function]}
    >
      Sign in with GitHub
    </button>
  </div>
</header>
`;
